edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/metrics/AvaliabilityMetrics.cs

```diff
@@ -90,40 +94,56 @@ public void ComputeAvailability(ModuleSet desired, ModuleSet current)
             /* Add new modules to track */
             foreach (string module in down.Union(up))
             {
-                this.availabilities.Add(new Availability(module, this.time));
+                this.availabilities.Add(new Availability(module, this.systemTime));
             }
         }
 
-        TimeSpan CalculateEdgeAgentDowntime()
+        public void Dispose()
+        {
+            this.updateCheckpointFile.Dispose();
+        }
+
+        /*
+         * The Logic below handles edgeAgent's own avaliability. It keeps a checkpoint file containing the current timestamp
+         *  that it updates every 5 minutes. On a clean shutdown, it deletes this file to indicate it shouldn't be running. If agent crashes
+         *  or returns a non-zero code, it leaves the file. On startup, if the file exists, agent knows it shutdown incorrectly \
+         *  and can calculate its downtime using the timestamp in the file.
+         */
+        public void IndicateCleanShutdown()
```

Comment:
Why not use AppDomain.CurrentDomain.ProcessExit?

Reply:
I only want this to happen on a clean shutdown. If it exits for some other reason, I don't want this to run. I'm pretty sure CurrentDomain.ProcessExit will run even if it crashed due to an unhanded exception.